src/MySqlConnector/MySqlClient/MySqlCommand.cs

```diff
@@ -111,7 +114,70 @@ protected override DbParameter CreateDbParameter()
 
 		internal async Task<int> ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
 		{
-			using (var reader = (MySqlDataReader) await ExecuteReaderAsync(CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false))
+			if (CommandType == CommandType.StoredProcedure)
+				return await ExecuteStoredProcedureAsync(ioBehavior, cancellationToken);
+
+			return await ExecuteTextCommandNonQueryAsync(CommandText, m_parameterCollection, ioBehavior, cancellationToken);
+		}
+
+		internal async Task<int> ExecuteStoredProcedureAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			var inParams = new MySqlParameterCollection();
+			var outParams = new MySqlParameterCollection();
+			var argParamNames = new List<string>();
+			var outParamNames = new List<string>();
+			MySqlParameter returnParameter = null;
+			for (var i = 0; i < m_parameterCollection.Count; i++)
+			{
+				var parameter = m_parameterCollection[i];
+				if (parameter.Direction == ParameterDirection.Input || parameter.Direction == ParameterDirection.InputOutput)
+				{
+					inParams.Add(parameter);
+				}
+				if (parameter.Direction == ParameterDirection.Output || parameter.Direction == ParameterDirection.InputOutput)
+				{
+					outParams.Add(parameter);
+					outParamNames.Add(parameter.ParameterName);
+				}
+
+				if (parameter.Direction == ParameterDirection.ReturnValue)
+					returnParameter = parameter;
+				else
+					argParamNames.Add(parameter.ParameterName);
+			}
+
+			int affectedRows;
+			var commandText = "CALL " + CommandText + "(" + string.Join(", ", argParamNames) +")";
+			using (var reader = (MySqlDataReader) await ExecuteTextCommandReaderAsync(commandText, inParams, CommandBehavior.SingleResult | CommandBehavior.SingleRow, ioBehavior, cancellationToken).ConfigureAwait(false))
+			{
+				do
+				{
+					if (await reader.ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false) && returnParameter != null)
+						returnParameter.Value = reader.GetValue(0);
+				} while (await reader.NextResultAsync(ioBehavior, cancellationToken).ConfigureAwait(false));
+				affectedRows = reader.RecordsAffected;
+			}
+
+			if (outParams.Count > 0)
+			{
+				commandText = "SELECT " + string.Join(", ", outParamNames);
+				using (var reader = (MySqlDataReader) await ExecuteTextCommandReaderAsync(commandText, new MySqlParameterCollection(), CommandBehavior.Default, ioBehavior, cancellationToken).ConfigureAwait(false))
+				{
+					await reader.ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+					for (var i = 0; i < outParams.Count; i++)
+					{
+						var parameter = outParams[i];
+						parameter.Value = reader.GetValue(i);
```

Comment:
Nitpick: unless you're going to use it for something else, I'd probably combine these into one line: `outParams[i].Value = reader.GetValue(i);`.